src/common/com/intellij/plugins/haxe/model/type/SpecificTypeReference.java

```diff
@@ -18,64 +18,106 @@
 package com.intellij.plugins.haxe.model.type;
 
 import com.intellij.psi.PsiElement;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 public abstract class SpecificTypeReference {
-  static public SpecificTypeReference ensure(SpecificTypeReference clazz) {
-    return (clazz != null) ? clazz : new SpecificHaxeClassReference(null, SpecificHaxeClassReference.EMPTY, null, null);
+  final protected PsiElement context;
+
+  public SpecificTypeReference(@NotNull PsiElement context) {
+    this.context = context;
+  }
+
+  static public SpecificTypeReference ensure(SpecificTypeReference clazz, @NotNull PsiElement context) {
```

Comment:
It ensures that the type is not null, providing a Dynamic in the case it is null.
It should receive a @Nullable and return a @NotNull. Also I can change the name to make it more readable. I liked it because it was just one simple word and I was potentially going to use it in lot of places.

In last commits I changed a lot of code to be @NotNull . In fact, ResultHolder is not mostly used without being null. I have found that lot of nulls caused a lot of bugs and null pointer exceptions. So I have tried to limit nulls as much as possible in latest commits.